new Go support function 'buildGoPackage'

ehmry opened the pull request:
This PR is for a new language-specific build funtion for Go.

I have some practical package expressions that use this fuction that I can append to the PR as well.
Also, I have a script that spits out basic expressions for building a Go package using only a Go import line, but I'm going to wait to do some more testing first.

This will take care of https://github.com/NixOS/nixpkgs/issues/3080.

lethalman commented on pkgs/development/go-modules/generic/default.nix:
```diff
@@ -0,0 +1,86 @@
+    else
+	find . -type d | while read d; do
+            for i in $d/*.go; do
+                go install -p $NIX_BUILD_CORES -v $d
```
This is executed for every .go file, why?

lethalman commented on pkgs/development/go-modules/generic/default.nix:
```diff
@@ -0,0 +1,86 @@
+
+    if [ -n "$subPackages" ] ; then
+	for p in $subPackages ; do
+            go install -p $NIX_BUILD_CORES -v $packagePath/$p
```
Shouldn't this be go test?

lethalman commented on pkgs/top-level/go-packages.nix:
```diff
@@ -0,0 +1,55 @@
+
+{ pkgs, overrides, stdenv }:
+
+let self = _self // overrides; _self = with self; {
```
This set must have also a `go` attribute, to know which go version are we using for this set of packages.

lethalman commented:
I'm willing to review this, @ehmry would you fix the couple of notes I've added to the changeset?

lethalman commented on pkgs/top-level/go-packages.nix:
```diff
@@ -0,0 +1,55 @@
+
+let self = _self // overrides; _self = with self; {
+
+  inherit (pkgs) buildGoPackage fetchgit fetchhg fetchurl;
```
These must be all inputs. Don't inherit from pkgs, thanks.

Commit:
new Go support function 'buildGoPackage'

add a setupHook to the Go compiler that builds a GOPATH
new buildGoPackage function that overides the phases of stdenv

ehmry commented:
@lethalman I've fixed the problems you pointed out but go.crypto doesn't build anymore, which breaks anything I wanted to do. Doesn't work with Go 1.3 or 1.4

The amount of work that would be involved in keeping buildGoPackage working for different package layouts and the quirks of go build/install is frightening me into considering calling Go too fragile and closing the PR.

All things considered I would prefer to build our own version of 'go build' that reads everything it needs from a builder environment and resolves and installs the way we like, but that not might be practical.

ehmry commented:
A lot of go build is exported as public interfaces, so it might not be so insane.
https://github.com/golang/go/tree/master/src/cmd/go

lethalman commented:
I'm sorry I can't say why it breaks crypto.

lethalman commented:
It's that I don't get why the changes I proposed hurt the build of crypto. Will have to check locally.

ehmry commented:
Bah

lethalman commented:
Btw the problem @ehmry seems more like fetchhg not being deterministic.

lethalman commented:
I was able to successfully build `goPackages.net`. @ehmry I don't understand the reaction, are you going to drop the whole work or perhaps open a new PR? Didn't get it.

lethalman commented:
I was even able to build `go13Packages.net`. This is a quick diff for having go14Packages and go13Packages:

```
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -7594,14 +7594,24 @@ let
 
   ### DEVELOPMENT / GO MODULES
 
-  buildGoPackage = import ../development/go-modules/generic {
-    inherit go;
-  };
+  go13Packages = recurseIntoAttrs (callPackage ./go-packages.nix {
+    go = go_1_3;
+    buildGoPackage = import ../development/go-modules/generic {
+      go = go_1_3;
+    };
+    overrides = (config.goPackageOverrides or (p: {})) pkgs;
+  });
 
-  goPackages = recurseIntoAttrs (callPackage ./go-packages.nix {
+  go14Packages = recurseIntoAttrs (callPackage ./go-packages.nix {
+    go = go_1_4;
+    buildGoPackage = import ../development/go-modules/generic {
+      go = go_1_4;
+    };
     overrides = (config.goPackageOverrides or (p: {})) pkgs;
   });
 
+  goPackages = go14Packages;
+
   ### DEVELOPMENT / LISP MODULES
 
   asdf = callPackage ../development/lisp-modules/asdf {
```

Will try this PR for a couple of real projects.

ehmry commented:
I'm going to drop the work, but if you want to take it on I'm willing to help.

lethalman commented:
Ok so I'm officially working on this. FYI @iElectric I'm working on porting ngrok to this new system.

iElectric commented:
:Beers:
